### DISCUSSION TITLE
[Question] Best practices ServiceClient & Dependency Injection

### DISCUSSION THREAD

**Discussion by mVermaat** [microsoft/PowerPlatform-DataverseServiceClient#312] [Q&A] [answered]
Do you have any guidance / best practices when using the ServiceClient in web applications or function apps? These will use dependency injection. Kind of similar to what was done for the [HttpClient](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests).

So basically: What is the correct way of using the ServiceClient with Dependency Injection in web applications?

**Comment by jordimontana82** [comment]
@mVermaat I think that is not coevered in the Dataverse docs (yet) but here's a nice example by Temmy @temmyraharjo:

https://temmyraharjo.wordpress.com/2022/07/09/how-to-make-azure-functions-dataverse-service-client/

**Comment by MattB-msft** [comment]
Yep.  this is on the "To-Do" list for samples for the client,  
also, we are internally testing a few different extensions for Dependency injection, one of which we may add to the client in the future.

**Comment by mVermaat** [comment]
Good to know, thanks for the information. With the current constraints, what is the best approach? I know I can make it 'work' by implementing something like in the link. However I'm not sure what the consequences are for using in example Scoped or Singleton.

**Comment by MattB-msft** [comment, marked as answer]
Scoped or Singleton uses cases are situational, IE, what are you trying to accomplish in the function or asp.net core app. 
We do not have a strong opinion on it and are using both processes in different services. 

Mainly what we are testing is memory and isolation management options for us in API services (function and the like), when running at scale over a long period of time.  Once we settle on a specific approach, we will build it into the client as an extension.  In the meantime, we may provide a sample or 2 on the PowerApps Samples site based on current work.

**Reply by jordimontana82** [reply]
@MattB-msft was wondering if some samples for this were published to the PowerApps Samples site in the end?

**Reply by KimBon** [reply]
@MattB-msft any update on this one? where can we see a latest sample or best practices for dependency injection?

**Comment by ferodom** [comment]
@MattB-msft  any updates I am having a hard time with the dependency injection taking about a second to create an instance of service client any suggestions?

this is what I am doing 

builder.Services.AddScoped<IOrganizationServiceAsync2>(x => 
    new ServiceClient(new ConnectionOptions()
    {
        ServiceUri = new Uri(datavserseUrl),
        AuthenticationType = AuthenticationType.ClientSecret,
        ClientId = clientId,
        ClientSecret = clientSecret,
    })
);

**Comment by FarmerenHE** [comment]
It works in both Singleton and Scoped setups, depending on your usecase it can be used differently but my recommendation would be to always use it as a singleton.

Scoped:
Pros:
Easy and quick to setup codewise.
Cons:
Takes a long time to create (often >300 ms)
Uses up connections from the connectionpool (primarily an issue with consumption based functions)

Singleton:
Cons:
You need to setup some handling to take care of the cases where the connection fails for one reason or another and then recreate the ServiceClient object and store it in the singleton.
Pros:
You dont need to spend time creating the serviceclient
It can be shared between function instances
Makes handling of API limits easier

If your running on consumptionbased azure functions you have a limit of 600 outbound connections, everytime you scope a new instances of the serviceclient it reserves one or more connections and they have to be unused for 5 minutes to be removed again. - thus making a singleton the only way to go in any sort of high load situation.

Personally I always use it as a singelton in my own custom build framework that ensures a connection is recreated if needed, handles the API limits and share the load between different connections if running in a high performance scenario. (We have tested it to 1400 calls/s sustained for an hour without running into API limits or azure connection limits).

**Reply by mVermaat** [reply]
@FarmerenHE This is a very useful insight. Thank you so much for sharing.

**Reply by mathiasbl** [reply]
You can use it as singleton if you use the Async Methods.
The sync methods have internal locking so your requests will be queued and processed one by one in a multithreaded environment.

It really depends on how you use the service client how you can register it.

You can always call .Clone() on the singleton serviceclient so you don't have the overhead of creating new connections.

**Reply by FarmerenHE** [reply]
If your using it in a non async way you should make sure to use a Clone anyway as you can run into issues with the connection hitting an error and needing to be recreated.

That said according to https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/447#issuecomment-2077838882 Clone and sync methods is being removed at some point and shouldn't be used in new code.

**Reply by KimBon** [reply]
> Personally I always use it as a singelton in my own custom build framework that ensures a connection is recreated if needed, handles the API limits and share the load between different connections if running in a high performance scenario. (We have tested it to 1400 calls/s sustained for an hour without running into API limits or azure connection limits).

@FarmerenHE what do you mean with sharing the load: parallel, multiple clients and users/clientids, ...?

**Reply by FarmerenHE** [reply]
> parallel, multiple clients and users/clientids, ...?

A bit of multiple practices. 

First and mostly I encurage the developers using the framework to write async azure functions and let azure handle how much load it should work with. (We wery often put stuff onto a servicebus and then have an azure function consume it, if we need a batchjob to run on a lot of data we typically make one job to query the IDs and put those on a servicebus and then have an azurefunction that handles a small batch of elements).

As for the framework itself it have its affinitycookie turned off (https://learn.microsoft.com/en-us/dotnet/api/microsoft.powerplatform.dataverse.client.serviceclient.enableaffinitycookie?view=dataverse-sdk-latest#microsoft-powerplatform-dataverse-client-serviceclient-enableaffinitycookie) and have some logic to invalidate it if its hit by API Limits.

We then have multiple sets of ClientId/Secrets that are used to query data to ensure that we always have an avaliable user that is ready to respond when we need to interact with Dataverse.

**Reply by KimBon** [reply]
Thanks already for answering my question!

Are your Azure Functions triggered by new service bus messages? If so how do you control the load and scale out as they can easily lead to _too many requests_ on the side of Dataverse? Ideally, if there is a 429 on Dataverse, new function instances should not be created and the running once should be able to finish in time (assuming consumption plan).

You turn affinity off by default is what you mean? Or you invalidate the cookie once there are issues?

We used to work with multiple identities in one API project, one per functional domain? That is similar to what you are doing?

**Reply by FarmerenHE** [reply]
We always turn affinity cookie off by default, most environments have 4+ dataverse servers, (and for any larger installation 9 is not unusual) - so just by that one change we have a lot more headroom in API limits, - it does give a potential slower response if we need to query the same data multiple times (or work with pages) but that is corner cases in our usages and we accept a performance penalty in those cases.

We usually find a upper limit of function scaleout where we can handle the usual load of requests, as well as keep it whitin the API limits, that have been enough for us to find a level of max scaling where there is some head room on both API limit and event handling capacity.

We do however try to limit our calls to dataverse as much as possible and make the queries as light as possible (by using queryexpressions and columnset wherever possible). - and the Context for using lambda expressions is a call hog that uses multiple calls for even simple things.

We have the same connectionstring with the same number of users between a large number of different functions, just to keep that part of the infrastructure simple, but yes you could use it that way if you wanted. - we just find that general API projects tend to grow over time and decrease performance to much compared to small lightweight functions with direct access.

**Reply by KimBon** [reply]
> You need to setup some handling to take care of the cases where the connection fails for one reason or another and then recreate the ServiceClient object and store it in the singleton.

@FarmerenHE how do you approach this? Wrapping service client in your own and provide your own methods checking the readiness of the client/connection?

> and the Context for using lambda expressions is a call hog that uses multiple calls for even simple things
You mean you are avoiding linq?

**Reply by FarmerenHE** [reply]
We have our own wrapper that extends IOrganizationServiceAsync2 and passes any calls to the internal ServiceClient (we also have a ton of extensions to make query expressions easier hidden in there).

When it hits API Limits (error 429) its own "IsReady" method invalidates it for the time that the service reports it to be invalid (using the Retry-After key from the errormessage).

Yeah were avoiding Linq alltogether and only using queryexpressions (or fetchxml in some cases) - XrmToolbox have a nice FetchXMLBuilder tool that allows to create the query and export it as QueryExpression, making development of complex queries quite simple.

**Reply by rappen** [reply]
You're welcome, @FarmerenHE 😉

**Reply by tpb1109** [reply]
> We have our own wrapper that extends IOrganizationServiceAsync2 and passes any calls to the internal ServiceClient (we also have a ton of extensions to make query expressions easier hidden in there).
> 
> When it hits API Limits (error 429) its own "IsReady" method invalidates it for the time that the service reports it to be invalid (using the Retry-After key from the errormessage).
> 
> Yeah were avoiding Linq alltogether and only using queryexpressions (or fetchxml in some cases) - XrmToolbox have a nice FetchXMLBuilder tool that allows to create the query and export it as QueryExpression, making development of complex queries quite simple.

Hey @FarmerenHE! I know this is an older thread, but I’d love to pick your brain on some approaches I’m taking for a large project for a web application that uses Dataverse heavily. I’m new to the azure development environment and some of the concepts of dependency injection, so I’d love to get your perspective on the way I’m handling certain things.

**Reply by FarmerenHE** [reply]
> Hey @FarmerenHE! I know this is an older thread, but I’d love to pick your brain on some approaches I’m taking for a large project for a web application that uses Dataverse heavily. I’m new to the azure development environment and some of the concepts of dependency injection, so I’d love to get your perspective on the way I’m handling certain things.

The short version is that you need to register dataverse as a singleton, have a way to handle multiple instances and preferably some sort of throttling functionality to ensure you dont hit API limits.

Your welcome to send me a message on linkedin - https://www.linkedin.com/in/simon-otto-nielsen-marqvarsen-46539722/ if you need something more specific

**Reply by tpb1109** [reply]
@FarmerenHE thanks! I sent you a connection request.

**Comment by khoait** [comment]
@FarmerenHE thanks for your valuable insights! are you also tweaking Max Concurrent Calls in the Azure Functions with Service Bus?

**Reply by FarmerenHE** [reply]
We have just left it running in more or less its default configuration, we usually raise the maximum allowed parallel instances instead (but no more than 2 pr. user as that more than that is usually the place were we start seeing API Limits anyway.

Were running azure functions in consumption plan as we find that have the best performance especially when we also look at the cost

**Reply by khoait** [reply]
thank you!

**Reply by NARESHARADHYULA** [reply]
@FarmerenHE 
What is recommendation for the lifetime scope of dataverse serviceclient when used in a WebApp that may run for months before being recycled?
How should the implementation be for a highvolume web application to avoid port exhaustion issues?

**Reply by FarmerenHE** [reply]
> @FarmerenHE What is recommendation for the lifetime scope of dataverse serviceclient when used in a WebApp that may run for months before being recycled? How should the implementation be for a highvolume web application to avoid port exhaustion issues?

In my initial framework I recycled it every 5 minutes, I have since changed that to a flow where it gets rebuild every time the code hits an exception from the call, the call is then tried again before the call is returned as faulty. (Its not perfect, but we generally get so few exceptions from our code that its good enough)

As for high volume make sure you never use .Clone() as that consumes a lot of ports, other than that I have found that if possible consumptionplan is giving the best performance and least port exhaustion issues.
